src/test/java/org/forgerock/cuppa/CuppaTest.java

```diff
@@ -299,6 +299,134 @@ public void basicApiUsageWithPassingFailingAndErroredTests() {
         assertTestResources(results, 1, 1, 1);
     }
 
+    @Test
+    public void beforeShouldRunOnceBeforeTests() {
+
+        //Given
+        AtomicInteger hasTopLevelBeforeFunctionRun = new AtomicInteger();
```

Comment:
I've gone off these `AtomicInteger`s. I think it's making the tests more confusing than they need to be. How about we create an implementation of our `Function` interface called `MockFunction`, which records the calls, like so:

```
class MockFunction implements Function {
    private int callCount = 0;

    public void apply() {
        callCount++;
    }

    public int getCallCount() { return callCount; }
}
```